extensions/identity-hub-verifier/src/main/java/org/eclipse/dataspaceconnector/identityhub/verifier/IdentityHubCredentialsVerifier.java

```diff
@@ -104,36 +104,70 @@ public Result<Map<String, Object>> getVerifiedCredentials(DidDocument didDocumen
 
         var verifiedCredentials = verifyCredentials(verifiableCredentials, didDocument);
 
-        monitor.debug(() -> String.format("Verified %s credentials", verifiedCredentials.size()));
+        monitor.debug(() -> String.format("Verified %s credentials", verifiedCredentials.getContent().size()));
 
-        var claims = extractClaimsFromCredential(verifiedCredentials);
+        var claims = extractClaimsFromCredential(verifiedCredentials.getContent());
 
-        if (claims.failed()) {
+        var failureMessages = Stream.concat(verifiedCredentials.getFailureMessages().stream(),
+                claims.getFailureMessages().stream()).collect(Collectors.toList());
+
+        var result = new VerificationResult<>(claims.getContent(), failureMessages);
+
+        if (result.failed()) {
             monitor.severe(() -> String.format("Credentials verification failed: %s", claims.getFailureDetail()));
-            return Result.failure(claims.getFailureDetail());
+            return Result.failure(result.getFailureDetail());
         } else {
-            return Result.success(claims.getContent());
+            return Result.success(result.getContent());
         }
     }
 
     @NotNull
-    private List<SignedJWT> verifyCredentials(StatusResult<Collection<SignedJWT>> jwts, DidDocument didDocument) {
-        var result = jwts.getContent()
+    private VerificationResult<List<SignedJWT>> verifyCredentials(StatusResult<Collection<SignedJWT>> jwts, DidDocument didDocument) {
+        // Get valid credentials.
+        var verifiedJwts = jwts.getContent()
+                .stream()
+                .map(jwt -> verifyJwtClaims(jwt, didDocument))
+                .collect(partitioningBy(AbstractResult::succeeded));
+
+        var jwtsSignedByIssuer = verifiedJwts.get(true)
+                .stream()
+                .collect(partitioningBy(jwt -> jwtCredentialsVerifier.isSignedByIssuer(jwt.getContent()).succeeded()));
+
+        var validCredentials = jwtsSignedByIssuer
+                .get(true)
+                .stream()
+                .map(AbstractResult::getContent)
+                .collect(Collectors.toList());
+
+        // Gather failure messages of invalid credentials.
+        var verificationFailures = verifiedJwts
+                .get(false)
                 .stream()
-                .collect(partitioningBy((jwt) -> jwtCredentialsVerifier.verifyClaims(jwt, didDocument.getId()) && jwtCredentialsVerifier.isSignedByIssuer(jwt)));
+                .map(AbstractResult::getFailureDetail);
 
-        var successfulResults = result.get(true);
-        var failedResults = result.get(false);
+        var signatureVerificationFailures = jwtsSignedByIssuer
+                .get(false)
+                .stream()
+                .map(AbstractResult::getFailureDetail);
+
+        var failedResults = Stream.concat(verificationFailures, signatureVerificationFailures)
```

Comment:
Consider adding a merge method to VerificationResult like [this one](https://github.com/agera-edc/DataSpaceConnector/blob/main/spi/core-spi/src/main/java/org/eclipse/dataspaceconnector/spi/result/Result.java#L56)